zilliqa/src/api/eth.rs

```diff
@@ -273,6 +276,119 @@ fn get_block_transaction_count_by_number(
+#[serde(default, rename_all = "camelCase")]
+struct GetLogsParams {
+    from_block: Option<BlockNumber>,
+    to_block: Option<BlockNumber>,
+    address: Option<OneOrMany<H160>>,
+    /// Topics matches a prefix of the list of topics from each log. An empty element slice matches any topic. Non-empty
+    /// elements represent an alternative that matches any of the contained topics.
+    ///
+    /// Examples (from Erigon):
+    /// * `[]`                          matches any topic list
+    /// * `[[A]]`                       matches topic A in first position
+    /// * `[[], [B]]` or `[None, [B]]`  matches any topic in first position AND B in second position
+    /// * `[[A], [B]]`                  matches topic A in first position AND B in second position
+    /// * `[[A, B], [C, D]]`            matches topic (A OR B) in first position AND (C OR D) in second position
+    topics: Vec<OneOrMany<H256>>,
+    block_hash: Option<H256>,
+}
+
+fn get_logs(params: Params, node: &Arc<Mutex<Node>>) -> Result<Vec<Log>> {
+    let params: GetLogsParams = params.one()?;
+
+    let node = node.lock().unwrap();
+
+    // Find the range of blocks we care about. This is an iterator of blocks.
+    let blocks = match (params.block_hash, params.from_block, params.to_block) {
+        (Some(block_hash), None, None) => Either::Left(std::iter::once(Ok(node
+            .get_block_by_hash(Hash(block_hash.0))?
+            .ok_or_else(|| anyhow!("block not found"))?))),
+        (None, from, to) => {
+            let from = node.get_view(from.unwrap_or(BlockNumber::Latest));
+            let to = node.get_view(to.unwrap_or(BlockNumber::Latest));
+
+            Either::Right((from..=to).map(|view| {
+                node.get_block_by_view(view)?
+                    .ok_or_else(|| anyhow!("missing block: {view}"))
+            }))
+        }
+        _ => {
+            return Err(anyhow!(
+                "only one of `blockHash` or (`fromBlock` and/or `toBlock`) are allowed"
+            ));
+        }
+    };
+
+    // Get the transaction hashes from each block. This is an iterator of (txn_index, txn_hash, block_number, block_hash).
+    let txn_hashes = blocks
+        .map(|block: Result<_>| {
+            let block = block?;
+            let block_number = block.view();
+            let block_hash = block.hash();
+            // TODO: Consider pre-filtering blocks at this point based on the log bloom filters.
+            Ok(block
+                .transactions
+                .into_iter()
+                .enumerate()
+                .map(move |(i, t)| (i, t, block_number, block_hash)))
+        })
+        .flatten_ok();
+
+    // Get the receipts for each transaction hash. This is an iterator of (receipt, txn_index, txn_hash, block_number, block_hash).
+    let receipts = txn_hashes.map(|h: Result<_>| {
+        let (txn_index, txn_hash, block_number, block_hash) = h?;
+        let receipt = node
+            .get_transaction_receipt(txn_hash)?
```

Comment:
Since I recently changed the way receipts are stored, I think it's likely worth  using `node.get_transaction_receipts_in_block()` rather than collecting txn_hashes and then individual receipts for each txn.

Reply:
Good idea, thanks for spotting.

Reply:
Ah one sticking point is I need to know the index of the transaction in the block, which `TransactionReceipt` currently doesn't contain. Do you think it makes sense to add that just for this API? On the one hand it is basically redundant data if you consider the `TransactionReceipt` tied to the `Transaction`. However, it does mean we can implement this slightly more efficiently...

Reply:
Ah yes I see the problem. Hmm - honestly I don't see any downside to including it in the TransactionReceipt. The index isn't stored on `Transaction`s either, and that means that any time we'd want it, we'd have to fetch the block, iterate through the block's tx hash list and find the position of the tx. Storing it in the receipt makes the most sense to me.

Reply:
Ah I found I'd also need to add the block's number to the receipts as well as the block hash. I'm wondering whether the extra complexity is worth it. The fact that we need to add this 'redundant' data to the receipts makes me feel un-easy like our data model is not quite right/natural.

@n-hutton It might be useful to get your opinion here too.

Reply:
Hm, what about querying the block number (using the hash) for every block in the RPC?
So if I understand it right, something along the lines of (in pseudocode)
```
let receipts: Vec<Vec<(Receipt, u64)>> = blocks.map(|block| {
  (block.get_transaction_receipts_in_block(), block.view())
});
let logs = receipts.map(|(receipt_vec, view)| {
  receipt_vec.map(|receipt| {
    receipt.logs.map(|log| { view, ...log })
  })
}).flatten().flatten()
// map logs to output format
```

Reply:
That makes sense - I was trying to avoid getting the block at all and just sticking with the block hash, but your suggestion makes more sense. Updated the PR, please take another look when you have a moment :)

Reply:
Ah I see! I think your previous commit was already fetching the blocks though which is why I suggested it this way.

Either way looks good now!